jax/_src/lax/control_flow/for_loop.py

```diff
@@ -229,9 +242,33 @@ def _get_ref_state_effects(jaxpr: core.Jaxpr) -> List[Set[StateEffect]]:
            if isinstance(eff, (ReadEffect, WriteEffect, AccumEffect))
            and eff.ref_aval is v.aval} for v in jaxpr.invars]
 
-@for_p.def_abstract_eval
+@for_p.def_effectful_abstract_eval
 def _for_abstract_eval(*avals, jaxpr, **__):
-  return list(avals)
+  # Find out for each of the `Ref`s in our jaxpr what effects they have.
+  jaxpr_aval_effects = _get_ref_state_effects(jaxpr)[1:]
+  aval_effects = [set(eff.replace(ref_aval=aval) for eff in effs) for aval, effs
+                  in zip(avals, jaxpr_aval_effects)
+                  if isinstance(aval, ShapedArrayRef)]
+  nonlocal_state_effects = core.join_effects(*aval_effects)
+  return list(avals), nonlocal_state_effects
+
+@state.register_discharge_rule(for_p)
+def _for_discharge_rule(in_avals, *args: Any, jaxpr: core.Jaxpr,
+                        reverse: bool, which_linear: Sequence[bool],
+                        nsteps: int
+                        ) -> Tuple[Sequence[Optional[Any]], Sequence[Any]]:
+  state_avals = [state.ShapedArrayRef(v.aval.shape, v.aval.dtype) for v in
+                 jaxpr.invars[1:]]
+  jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(
```

Comment:
Maybe we don't need to make a new jaxpr! We're basically substituting ref-valued arguments with non-ref-valued arguments, but the `for_loop` itself (including its body jaxpr) can stay the same.